### ISSUE TITLE
Support signing multiple artifacts with a single certificate

### ISSUE THREAD

**Issue by mayaCostantini** [sigstore/sigstore-python#514]
**Description**

For signing a large number of input files, supporting signing multiple artifacts with a single certificate (and eventually asynchronous signing) in `sigstore-python` could allow saving time and API calls to Fulcio and OIDC providers. Instead of creating a new key pair and requesting a different signing certificate from Fulcio when signing every artifact, those two first steps could be shared for a group of artifacts, and the rest of the steps executed sequentially or asynchronously.

This would result in the following flow when multiple artifacts are passed to sigstore to be signed:

- Generate a single key pair
- Record the current timestamp
- Get an identity token from an OIDC provider
- Request a single signing certificate from Fulcio using the id token
- Verify the certificate and sign off the artifacts (asynchronously)
- When signing, verify that the current certificate is not expired by comparing it to the recorded timestamp. If the validity time of the certificate is exceeded, request a new certificate.
- Send (asynchronous) requests to Rekor to create a signature entry for every artifact

This feature could be configurable to allow users to turn it off and request one certificate per artifact if needed (even though I am not able to think of any reason for having a different certificate for each artifact).

If the asynchronous part turns out to be too complex and involves rewriting too many parts of the library, we could simply use the current sequential signing flow for all artifacts, keeping the shared steps (key generation and certificate request) as the only optimizations.

cc @lukehinds @znewman01

**Comment by woodruffw** [comment]
I might be wrong, but I believe we already do this: our `Signer` instance does the OIDC and CSR flows once per instance, and allows multiple artifacts to be signed against that instance.

**Comment by woodruffw** [comment]
Whoops, I did indeed get this slightly wrong: `sigstore sign` only does the OIDC identity token flow once per invocation, but the Signer itself *always* performs key generation + CSR for each individual signing operation. So we do indeed spend quite a bit of time on network trips that could be flattened down to a single key and cert.

**Comment by mayaCostantini** [comment]
@woodruffw I would be interested on working on this issue if no one else is assigned to it.

**Comment by woodruffw** [comment]
> @woodruffw I would be interested on working on this issue if no one else is assigned to it.

Yeah, please do! I'd be open to this being a new kwarg on the `Signer` constructor, which would then perform the separate "single" flow you've described.

**Comment by mayaCostantini** [comment]
Sounds good, thanks for the suggestions :slightly_smiling_face:

**Comment by mayaCostantini** [comment]
Thanks for the reminder and sorry for the delay, I opened a PR here: https://github.com/sigstore/sigstore-python/pull/645

**Comment by jku** [comment]
Could someone explain what is the use case for separate CSRs and keys? Who wants that?

I had no idea multiple artifacts currently mean multiple CSRs (and in the PR that still seems to be the default). It sounds quite unappealing to me.

**Comment by woodruffw** [comment]
> Could someone explain what is the use case for separate CSRs and keys? Who wants that?

Conceptually, nobody should want or not want it: the CSR and key shouldn't matter at all, since a Sigstore "identity" is whatever set of certificates are bound to the original OIDC (or other identity source) credential. 

The main reason we implemented it this way is because it's simpler: the `Signer` class is almost entirely stateless between signing operations, with no need to retain or check the live-ness of the key material it generates and requests. This reduces the possible space for errors that users can't directly resolve (e.g. creating a `Signer`, signing one thing, waiting a bit, and then discovering that their `Signer` has silently become invalid).

That being said, I think there's a legitimate performance and reliability argument for doing the CSR and key generation flow as few times as possible, and even for making that the default (for those reasons). It just shouldn't affect the actual *identity properties* of the Sigstore signatures, either way 🙂
